# Day 4: Pandas Fundamentals

In this notebook, I will learn how to work with tabular data using Pandas. I will start with Series and DataFrames, then use a real CSV dataset to inspect, clean, filter, and group data.

## Importing Pandas

Pandas is a Python library used for working with tabular data. It is commonly imported using the alias `pd`.

In [1]:
import pandas as pd

print(pd.__version__)

3.0.3


## Pandas Series

A Series is a one-dimensional collection of values with an index. It is similar to one labeled column in a spreadsheet.

In [2]:
scores = pd.Series([78, 85, 91, 88])
scores

0    78
1    85
2    91
3    88
dtype: int64

### Series Data Type

A Series has a data type that describes the kind of values stored inside it. Pandas usually detects the data type automatically.

In [3]:
integer_values = pd.Series([100, 102, 104])
decimal_values = pd.Series([100.0, 102.5, 104.2])
text_values = pd.Series(["A", "B", "C"])
boolean_values = pd.Series([True, False, True])

print("Integer dtype:", integer_values.dtype)
print("Decimal dtype:", decimal_values.dtype)
print("Text dtype:", text_values.dtype)
print("Boolean dtype:", boolean_values.dtype)

Integer dtype: int64
Decimal dtype: float64
Text dtype: str
Boolean dtype: bool


### Custom Index Labels

Pandas creates a numeric index by default, but custom labels can make the Series easier to understand and access.

In [4]:
student_scores = pd.Series(
    [100, 85, 91, 66],
    index=["Ahmad", "Mohammad", "Omar", "Sami"]
)

student_scores

Ahmad       100
Mohammad     85
Omar         91
Sami         66
dtype: int64

### Selecting by Label with loc

The `.loc` property selects a value using its index label.

In [5]:
print(student_scores.loc["Ahmad"])

100


### Selecting by Position with iloc

The `.iloc` property selects a value using its integer position.

In [6]:
print(student_scores.iloc[0])

100


### Comparing loc and iloc

Both `.loc` and `.iloc` can select values from a Series, but `.loc` uses labels while `.iloc` uses integer positions.

In [7]:
print("Using loc:", student_scores.loc["Omar"])
print("Using iloc:", student_scores.iloc[2])

Using loc: 91
Using iloc: 91


### Updating a Series Value

A value can be changed by selecting its label with `.loc` and assigning a new value.

In [8]:
student_scores.loc["Omar"] = 94

student_scores

Ahmad       100
Mohammad     85
Omar         94
Sami         66
dtype: int64

### Changing a Value Using Its Current Value

An existing value can also be increased or decreased instead of being completely replaced.

In [9]:
student_scores.loc["Ahmad"] += 2
student_scores

Ahmad       102
Mohammad     85
Omar         94
Sami         66
dtype: int64

### Filtering a Series

Filtering keeps only the values that match a condition. The condition first creates True and False values for every item in the Series.

In [10]:
high_score_mask = student_scores >= 85

high_score_mask

Ahmad        True
Mohammad     True
Omar         True
Sami        False
dtype: bool

In [11]:
high_scores = student_scores[high_score_mask]

high_scores

Ahmad       102
Mohammad     85
Omar         94
dtype: int64

### More Filtering Conditions

Different comparison operators can be used to filter a Series.

In [12]:
scores_below_85 = student_scores[student_scores < 85]
scores_above_90 = student_scores[student_scores > 90]

print("Scores below 85:")
print(scores_below_85)

print("\nScores above 90:")
print(scores_above_90)

Scores below 85:
Sami    66
dtype: int64

Scores above 90:
Ahmad    102
Omar      94
dtype: int64


### Creating a Series from a Dictionary

A Series can also be created from a Python dictionary. The dictionary keys become the index labels, and the dictionary values become the Series values.

In [13]:
study_hours_data = {
    "Sunday": 2,
    "Monday": 3,
    "Tuesday": 1.5,
    "Wednesday": 4
}

study_hours = pd.Series(study_hours_data)

study_hours

Sunday       2.0
Monday       3.0
Tuesday      1.5
Wednesday    4.0
dtype: float64

In [14]:
print("Monday study hours:", study_hours.loc["Monday"])

long_study_days = study_hours[study_hours >= 3]

long_study_days

Monday study hours: 3.0


Monday       3.0
Wednesday    4.0
dtype: float64

### Series Summary

In this section, I created Series objects from lists and dictionaries. I used custom index labels, selected values with `.loc` and `.iloc`, updated values, and filtered the Series using conditions.

## Pandas DataFrames

A DataFrame is a two-dimensional table with rows and named columns. It is similar to a spreadsheet and can store different types of data in different columns.

In [15]:
employee_data = {
    "name": ["Ahmad", "Lina", "Omar"],
    "age": [21, 23, 25],
    "department": ["IT", "Design", "Finance"]
}

employees_df = pd.DataFrame(employee_data)

employees_df

,name,age,department
0,Ahmad,21,IT
1,Lina,23,Design
2,Omar,25,Finance


### Rows, Columns, and Index

Each row represents one record, while each column represents one type of information. Pandas also creates an index for the rows.

In [16]:
employees_df

,name,age,department
0,Ahmad,21,IT
1,Lina,23,Design
2,Omar,25,Finance


### Custom Row Labels

A DataFrame uses numeric row labels by default, but custom labels can be provided when the DataFrame is created.

In [17]:
employees_df = pd.DataFrame(
    employee_data,
    index=["employee_1", "employee_2", "employee_3"]
)

employees_df

,name,age,department
employee_1,Ahmad,21,IT
employee_2,Lina,23,Design
employee_3,Omar,25,Finance


### Selecting Rows with loc

The `.loc` property selects a row using its index label.

In [18]:
employees_df.loc["employee_2"]

name            Lina
age               23
department    Design
Name: employee_2, dtype: object

In [19]:
print(type(employees_df))
print(type(employees_df.loc["employee_2"]))

<class 'pandas.DataFrame'>
<class 'pandas.Series'>


### Selecting Rows with iloc

The `.iloc` property selects a row using its integer position.

In [20]:
employees_df.iloc[0]

name          Ahmad
age              21
department       IT
Name: employee_1, dtype: object

### Comparing loc and iloc

`.loc` selects by row label, while `.iloc` selects by integer position.

In [21]:
print(employees_df.loc["employee_1"])
print("\n")
print(employees_df.iloc[0])

name          Ahmad
age              21
department       IT
Name: employee_1, dtype: object


name          Ahmad
age              21
department       IT
Name: employee_1, dtype: object


### Adding a New Column

A new column can be added by assigning a list of values to a new column name. The number of values must match the number of rows.

In [22]:
employees_df["salary"] = [900, 1100, 1300]

employees_df

,name,age,department,salary
employee_1,Ahmad,21,IT,900
employee_2,Lina,23,Design,1100
employee_3,Omar,25,Finance,1300


In [23]:
employees_df["active"] = True

employees_df

,name,age,department,salary,active
employee_1,Ahmad,21,IT,900,True
employee_2,Lina,23,Design,1100,True
employee_3,Omar,25,Finance,1300,True


### Adding a New Row

A new row can be created as a small DataFrame and then joined to the existing DataFrame using `pd.concat()`.

In [24]:
new_employee=pd.DataFrame([{"name":"Rami",
                            "age": 19,
                             "department":"IT",
                              "salary":950,
                               "active":True}],index=["employee_4"])
print(new_employee)
print()
df=pd.concat([employees_df,new_employee])
print(df)

            name  age department  salary  active
employee_4  Rami   19         IT     950    True

             name  age department  salary  active
employee_1  Ahmad   21         IT     900    True
employee_2   Lina   23     Design    1100    True
employee_3   Omar   25    Finance    1300    True
employee_4   Rami   19         IT     950    True


In [25]:
new_employees = pd.DataFrame(
    [{"name": "Yousef","age": 24,"department": "IT","salary": 1200,"active": True},
     {"name": "Maya","age": 26,"department": "Finance","salary": 1400,"active": True}],
     index=["employee_5", "employee_6"]
)
employees_df = pd.concat([employees_df, new_employees])

### Checking the DataFrame

After adding columns and rows, I displayed the DataFrame again to make sure the new data was added correctly.

In [26]:
employees_df

,name,age,department,salary,active
employee_1,Ahmad,21,IT,900,True
employee_2,Lina,23,Design,1100,True
employee_3,Omar,25,Finance,1300,True
employee_5,Yousef,24,IT,1200,True
employee_6,Maya,26,Finance,1400,True


### Adding New Rows with concat

I created another dictionary for the new employees, converted it into a DataFrame, and combined it with the original DataFrame using `pd.concat()`.

In [27]:
new_employee_data = {
    "name": ["Rami", "Sara"],
    "age": [19, 22],
    "department": ["IT", "Marketing"],
    "salary": [950, 1000],
    "active": [True, True]
}

new_employees_df = pd.DataFrame(
    new_employee_data,
    index=["employee_4", "employee_5"]
)

employees_df = pd.concat([employees_df, new_employees_df])

employees_df

,name,age,department,salary,active
employee_1,Ahmad,21,IT,900,True
employee_2,Lina,23,Design,1100,True
employee_3,Omar,25,Finance,1300,True
employee_5,Yousef,24,IT,1200,True
employee_6,Maya,26,Finance,1400,True
employee_4,Rami,19,IT,950,True
employee_5,Sara,22,Marketing,1000,True


### DataFrame Summary



In this section, I created a DataFrame from a dictionary and used custom row labels. I selected rows using `.loc` and `.iloc`, added new columns, and added new rows using `pd.concat()`.

## Loading the Dataset

I chose the Titanic passenger dataset for this exercise. It contains passenger information such as age, sex, ticket class, fare, and survival status.

The CSV file is stored in the same folder as this notebook, so I loaded it using a relative path.

In [28]:
titanic_df = pd.read_csv("train.csv")

## Inspecting the Dataset

Before cleaning the data, I checked its rows, columns, shape, and data types. This helped me understand the dataset before making any changes.

In [29]:
titanic_df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [30]:
titanic_df.tail()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.00,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.00,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.45,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.00,C148,C
890,891,0,3,"Dooley, Mr. Patrick",male,32.0,0,0,370376,7.75,NaN,Q


In [31]:
titanic_df.sample(5)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
752,753,0,3,"Vande Velde, Mr. Johannes Joseph",male,33.0,0,0,345780,9.5000,NaN,S
618,619,1,2,"Becker, Miss. Marion Louise",female,4.0,2,1,230136,39.0000,F4,S
417,418,1,2,"Silven, Miss. Lyyli Karoliina",female,18.0,0,2,250652,13.0000,NaN,S
205,206,0,3,"Strom, Miss. Telma Matilda",female,2.0,0,1,347054,10.4625,G6,S
819,820,0,3,"Skoog, Master. Karl Thorsten",male,10.0,3,2,347088,27.9000,NaN,S


In [32]:
print("Dataset shape:", titanic_df.shape)

Dataset shape: (891, 12)


In [33]:
print("Column names:")
print(titanic_df.columns.tolist())

Column names:
['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked']


In [34]:
print("Data types:")
print(titanic_df.dtypes)

Data types:
PassengerId      int64
Survived         int64
Pclass           int64
Name               str
Sex                str
Age            float64
SibSp            int64
Parch            int64
Ticket             str
Fare           float64
Cabin              str
Embarked           str
dtype: object


In [35]:
titanic_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 83.7 KB


In [36]:
titanic_df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


## Selecting Data

I selected specific columns and rows from the dataset instead of displaying the full table. This makes it easier to focus on the information I need.

In [37]:
titanic_df["Name"]

0                                Braund, Mr. Owen Harris
1      Cumings, Mrs. John Bradley (Florence Briggs Th...
2                                 Heikkinen, Miss. Laina
3           Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                               Allen, Mr. William Henry
                             ...                        
886                                Montvila, Rev. Juozas
887                         Graham, Miss. Margaret Edith
888             Johnston, Miss. Catherine Helen "Carrie"
889                                Behr, Mr. Karl Howell
890                                  Dooley, Mr. Patrick
Name: Name, Length: 891, dtype: str

### Selecting Columns

Selecting one column returns a Series, while selecting multiple columns returns a smaller DataFrame.

In [38]:
 titanic_df[["Name", "Sex", "Age", "Pclass", "Fare"]].head()



,Name,Sex,Age,Pclass,Fare
0,"Braund, Mr. Owen Harris",male,22.0,3,7.2500
1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,71.2833
2,"Heikkinen, Miss. Laina",female,26.0,3,7.9250
3,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,53.1000
4,"Allen, Mr. William Henry",male,35.0,3,8.0500


### Using a Column as the Index

I loaded another copy of the dataset using `PassengerId` as the row index. This makes it possible to select passengers by their ID using `.loc`.

In [39]:
passengers_by_id = pd.read_csv("train.csv",index_col="PassengerId")
passengers_by_id.head()

,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
PassengerId,,,,,,,,,,,
1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


### Selecting Rows with loc

The `.loc` property selects rows using their index labels.

In [40]:
passengers_by_id.loc[1]

Survived                          0
Pclass                            3
Name        Braund, Mr. Owen Harris
Sex                            male
Age                            22.0
SibSp                             1
Parch                             0
Ticket                    A/5 21171
Fare                           7.25
Cabin                           NaN
Embarked                          S
Name: 1, dtype: object

In [41]:
passengers_by_id.loc[1,["Name", "Sex", "Age", "Survived"]]

Name        Braund, Mr. Owen Harris
Sex                            male
Age                            22.0
Survived                          0
Name: 1, dtype: object

In [42]:
passengers_by_id.loc[1:5,["Name", "Age", "Fare"]]

,Name,Age,Fare
PassengerId,,,
1,"Braund, Mr. Owen Harris",22.0,7.2500
2,"Cumings, Mrs. John Bradley (Florence Briggs Th...",38.0,71.2833
3,"Heikkinen, Miss. Laina",26.0,7.9250
4,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",35.0,53.1000
5,"Allen, Mr. William Henry",35.0,8.0500


### Selecting Rows with iloc

The `.iloc` property selects rows and columns using their integer positions.

In [43]:
passengers_by_id.iloc[0:5, 0:4]

,Survived,Pclass,Name,Sex
PassengerId,,,,
1,0,3,"Braund, Mr. Owen Harris",male
2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female
3,1,3,"Heikkinen, Miss. Laina",female
4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female
5,0,3,"Allen, Mr. William Henry",male


### Selection Summary

In this section, I selected single and multiple columns from the dataset. I also used `.loc` to select by index labels and `.iloc` to select by integer positions.

## Filtering the Dataset

Filtering keeps only the rows that match a condition. I started with a simple condition, then combined two conditions to create a more specific subset.

In [44]:
adult_passengers = titanic_df[titanic_df["Age"] >= 18]

adult_passengers[["Name", "Age", "Sex", "Pclass"]].head()

,Name,Age,Sex,Pclass
0,"Braund, Mr. Owen Harris",22.0,male,3
1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",38.0,female,1
2,"Heikkinen, Miss. Laina",26.0,female,3
3,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",35.0,female,1
4,"Allen, Mr. William Henry",35.0,male,3


### Female Passengers Who Survived

I filtered the dataset to find female passengers who survived. I used `&` because both conditions must be true.

In [45]:
female_survivors=titanic_df[(titanic_df["Sex"] == "female")&(titanic_df["Survived"] == 1)]
female_survivors[["Name", "Age", "Pclass", "Fare"]].head(10)

,Name,Age,Pclass,Fare
1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",38.0,1,71.2833
2,"Heikkinen, Miss. Laina",26.0,3,7.9250
3,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",35.0,1,53.1000
8,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",27.0,3,11.1333
9,"Nasser, Mrs. Nicholas (Adele Achem)",14.0,2,30.0708
10,"Sandstrom, Miss. Marguerite Rut",4.0,3,16.7000
11,"Bonnell, Miss. Elizabeth",58.0,1,26.5500
15,"Hewlett, Mrs. (Mary D Kingcome)",55.0,2,16.0000
19,"Masselmani, Mrs. Fatima",NaN,3,7.2250
22,"McGowan, Miss. Anna ""Annie""",15.0,3,8.0292


In [46]:
female_survivor_count = female_survivors.shape[0]
print("Number of female survivors:", female_survivor_count)

Number of female survivors: 233


The filtered result contains 233 female passengers who survived. They came from different ages and passenger classes.

## Aggregation

Aggregation functions summarize many values into one result. I used them to calculate simple statistics from numeric columns in the dataset.

In [47]:
print("Average age:", titanic_df["Age"].mean())
print("Average fare:", titanic_df["Fare"].mean())
print("Minimum fare:", titanic_df["Fare"].min())
print("Maximum fare:", titanic_df["Fare"].max())
print("Number of known ages:", titanic_df["Age"].count())

Average age: 29.69911764705882
Average fare: 32.204207968574636
Minimum fare: 0.0
Maximum fare: 512.3292
Number of known ages: 714


## Grouping by Passenger Class

I grouped the passengers by ticket class. For each class, I calculated the average survival value, average fare, and number of passengers.

In [48]:
group = titanic_df.groupby("Pclass")
print(group["Survived"].mean())
print()
print(group["Fare"].mean())
print()
print(group["PassengerId"].count())


Pclass
1    0.629630
2    0.472826
3    0.242363
Name: Survived, dtype: float64

Pclass
1    84.154687
2    20.662183
3    13.675550
Name: Fare, dtype: float64

Pclass
1    216
2    184
3    491
Name: PassengerId, dtype: int64


### Groupby Result

The result shows that first-class passengers had the highest survival rate and paid the highest average fare. Third-class passengers had the lowest survival rate and also formed the largest passenger group.

## Data Cleaning

Before changing the dataset, I checked the number of missing values in each column.

In [49]:
titanic_df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [50]:
clean_titanic_df = titanic_df.copy()
age_median = clean_titanic_df["Age"].median()
clean_titanic_df["Age"] = clean_titanic_df["Age"].fillna(age_median)
most_common_port = clean_titanic_df["Embarked"].mode()[0]
clean_titanic_df["Embarked"] = clean_titanic_df["Embarked"].fillna(most_common_port)
clean_titanic_df = clean_titanic_df.drop(columns=["Cabin"])
clean_titanic_df.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64

### Missing Values Decision

I filled the missing `Age` values with the median because the median is less affected by unusually high or low values. I filled the missing `Embarked` values with the most common port because it is a categorical column. I removed the `Cabin` column because most of its values were missing.

### Duplicate Rows

I checked whether the dataset contained fully duplicated rows.

In [51]:
duplicate_count = clean_titanic_df.duplicated().sum()
print("Duplicate rows:", duplicate_count)
clean_titanic_df = clean_titanic_df.drop_duplicates()

Duplicate rows: 0


No fully duplicated rows were found, so removing duplicates did not change the dataset.

### Data Types

I checked the data types after cleaning to see whether any columns needed conversion.

In [52]:
clean_titanic_df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name               str
Sex                str
Age            float64
SibSp            int64
Parch            int64
Ticket             str
Fare           float64
Embarked           str
dtype: object

The columns were already stored using suitable data types, so no type conversion was needed.